### ISSUE TITLE
Inquiry Regarding the Interpretation of YOLO11 Hyperparameter Tuning Results

### ISSUE THREAD

**Issue by bjh03205** [ultralytics/ultralytics#20187]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I have a few questions regarding the interpretation of the hyperparameter tuning results for the YOLO11 model.

I have two specific inquiries:

Upon reviewing the tune_scatter_plots.png, I noticed that 7 of the 23 hyperparameters are set to a default value of 0. The absence of any variance during the random search tuning process suggests that these parameters are fixed. Could you please confirm if these values are intentionally configured to remain unchanged?

If these hyperparameters are indeed set to be fixed, I am curious to know the reasoning behind this decision.

I have also attached an example result for your reference.

![Image](https://github.com/user-attachments/assets/a147cb21-4c2f-46c4-98fa-bf5f672be378)
<Some results were artificially masked>

Thank you very much for your time and assistance.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @bjh03205, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com/) for additional resources on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, as many common questions are addressed there.

This is an automated response to help streamline your experience. An Ultralytics engineer will also review your question and assist you further soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate.

For custom training or hyperparameter tuning ❓ Questions, please share as much detail as possible (including dataset samples, training logs, and config files) and confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community in your preferred space! For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For deeper discussions, try [Discourse](https://community.ultralytics.com/), or join our [Subreddit](https://reddit.com/r/Ultralytics) to connect with other users.

## Upgrade

To make sure your issue isn't already resolved, upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of the following up-to-date verified environments (with all dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your detailed question and for providing reference images! Your inquiry has been received and will be reviewed by the team soon.

**Comment by Y-T-G** [comment]
What was the command you used to launch tuning?

**Comment by bjh03205** [comment]
Most settings were configured with default values, and the training command is as follows.

```python
from ultralytics import YOLO

model = YOLO('yolo11s.yaml').load()
model.tune(data="training.yaml", epochs=30, iterations=300, optimizer='SGD', plots=False, save=False, val=False, device=1)
```

**Comment by bjh03205** [comment]
Thank you very much for your kind and detailed response.

Lastly, I am curious to know what specific reasons led to intentionally deactivating these seven parameters(degrees, shear, perspective, flipud, bgr, mixup, copy_paste) out of the 23 total hyperparameters.

For instance, was this decision made because tuning these parameters was considered relatively less effective in enhancing the model's performance than others?

**Comment by glenn-jocher** [comment]
The specific choice of which parameters to deactivate by default is based on empirical results and the nature of the augmentations themselves. These particular seven parameters are initially set to zero for good reasons:

1. Some create unrealistic scenarios for many datasets - `flipud` (vertical flipping) rarely reflects real-world conditions since objects typically maintain consistent up-down orientation.

2. Others like `shear`, `perspective`, and `degrees` can introduce significant distortions that might negatively impact model performance unless your dataset specifically benefits from these transformations.

3. `mixup`, `copy_paste`, and `bgr` are more specialized augmentations that tend to be beneficial only for specific use cases rather than general object detection scenarios.

These default values represent a conservative augmentation strategy that works well across most datasets. For specific domains where these transformations would better represent real-world variation, you can explicitly include them in your custom search space as shown in my previous response.

**Comment by bjh03205** [comment]
Thank you very much for your prompt and professional response :)

**Comment by glenn-jocher** [comment]
You're very welcome! Happy to help with your hyperparameter tuning questions. Feel free to reach out if you have any other queries about Ultralytics models in the future.

**Comment by anayyyxx** [comment]
> The behavior you're observing is expected with the Ultralytics tuning implementation. Hyperparameters that are initialized to 0 in the search space are effectively excluded from the tuning process. This is an intentional design choice to help focus the search on the most impactful parameters while keeping others fixed.
> 
> When a hyperparameter value is set to 0 initially (like some of your augmentation parameters), the tuner doesn't mutate these values during the search process. This approach serves several purposes:
> 
> 1. It reduces the dimensionality of the search space, making the tuning process more efficient
> 2. It allows you to deliberately fix certain parameters while exploring others
> 3. For augmentation parameters specifically, a value of 0 often means "disabled," which can be desirable in certain scenarios
> 
> If you want to include these parameters in your search, you can define a custom search space:
> 
> custom_space = {
>     "lr0": (1e-5, 1e-1),
>     "degrees": (0.0, 45.0),  # Including degrees in search
>     "shear": (0.0, 10.0),    # Including shear in search
>     # Add other parameters you want to tune
> }
> 
> model.tune(data="training.yaml", epochs=30, iterations=300, space=custom_space)
> This will ensure all parameters in your custom space are tuned regardless of their initial values.

Even after specifying a custom space for these hyperparameters, it's still kept as a constant 0

**Comment by glenn-jocher** [comment]
It sounds like you're experiencing an issue with the custom search space implementation. When using the native Ultralytics tuner (not Ray Tune), the search space format is different - it should be a dictionary where each key maps to a tuple of (min, max) or (min, max, gain).

Try this format instead:

```python
custom_space = {
    "lr0": (1e-5, 1e-1),
    "degrees": (0.0, 45.0),  # Will now be tuned
    "shear": (0.0, 10.0),    # Will now be tuned
    # Other parameters
}

model.tune(data="training.yaml", epochs=30, iterations=300, space=custom_space)
```

If you're using Ray Tune (with `use_ray=True`), then your space should use the Ray Tune API format:

```python
from ray import tune

custom_space = {
    "lr0": tune.uniform(1e-5, 1e-1),
    "degrees": tune.uniform(0.0, 45.0),
    "shear": tune.uniform(0.0, 10.0),
}

model.tune(data="training.yaml", use_ray=True, space=custom_space)
```

Make sure you're using the correct format based on which tuning method you're employing.

**Comment by javierselva** [comment]
> Even after specifying a custom space for these hyperparameters, it's still kept as a constant 0

I am facing this same issue. I think the problem is that, as stated above, the model takes a previous iteration and mutates it. The initial configuration is marked by the parameters you pass as "initial configuration", the same as you would do when training a model a single time. If they're not specified, it will use the default parameters. You can actually see that [the ones that are set to 0 by default](https://docs.ultralytics.com/modes/train/#augmentation-settings-and-hyperparameters) are the ones which remained fixed in OPs post. 

The problem is, as the mutations are computed with a multiplicative factor, if the parameter was 0 (or close to 0 in my experience) it effectively remains as 0 for the whole hyperparameter search, regardless of the search space:

https://github.com/ultralytics/ultralytics/blob/5cf31ada2dcf7b8ec474f18e2849618855c8cdc1/ultralytics/engine/tuner.py#L146

So, if you want them to mutate, aside from including them in your search space, you should initialize them with some >0 values. For instance:
```
custom_space = {
    "degrees": (0.0, 45.0), 
    "perspective": (0.0, 1.0),
    "flipud": (0.0, 1.0),
    "bgr": (0.0, 1.0),
    "mixup": (0.0, 1.0),
    "copy_paste": (0.0, 1.0),
    "shear": (0.0, 10.0), 
    # Other parameters ...
}

model.tune(..., space=custom_space, degrees=20, shear=0.1, flipud=0.5, perspective=0.5, bgr=0.5, mixup=0.5, copy_paste=0.5, ...)
```

Note that very small values do not seem to do the trick, I tested with 1e-5 and still no change.

**Comment by glenn-jocher** [comment]
@javierselva has provided an excellent analysis of the root cause - you're absolutely correct that parameters initialized to 0 remain fixed during evolutionary tuning due to the multiplicative mutation approach in the codebase. This is indeed a limitation of the current implementation.

Your workaround of setting initial non-zero values alongside the custom search space is the correct solution. Thank you for the detailed investigation and code reference - this helps explain why simply defining a custom search space wasn't sufficient for @anayyyxx.

This behavior could certainly be improved in future versions to handle zero-initialized parameters more intelligently during the mutation process.
